### PR TITLE
DOC: update the aggregate docstring

### PR DESCRIPTION
**Pull request by albertvillanova**
Checklist for the pandas documentation sprint (ignore this if you are doing
an unrelated PR):

- [X] PR title is "DOC: update the <your-function-or-method> docstring"
- [ ] The validation script passes: `scripts/validate_docstrings.py <your-function-or-method>`
- [X] The PEP8 style check passes: `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [X] The html version looks good: `python doc/make.py --single <your-function-or-method>`
- [ ] It has been proofread on language by another sprint participant

Please include the output of the validation script below between the "```" ticks:

```
################################################################################
#################### Docstring (pandas.DataFrame.aggregate) ####################
################################################################################

Aggregate using one or multiple operations along the specified axis.

.. versionadded:: 0.20.0

Parameters
----------
func : function, string, dictionary, or list of string/functions
    Function to use for aggregating the data. If a function, must either
    work when passed a DataFrame or when passed to DataFrame.apply. For
    a DataFrame, can pass a dict, if the keys are DataFrame column names.

    Accepted combinations are:

    - string function name.
    - function.
    - list of functions.
    - dict of column names -> functions (or list of functions).
axis : {0 or 'index', 1 or 'columns'}, default 0
    - 0 or 'index': apply function to each column.
    - 1 or 'columns': apply function to each row.
args
    Optional positional arguments to pass to the function.
kwargs
    Optional keyword arguments to pass to the function.

Returns
-------
aggregated : DataFrame

Notes
-----
`agg` is an alias for `aggregate`. Use the alias.

Notes
-----
The default behavior of aggregating over the axis 0 is different from
`numpy` functions `mean`/`median`/`prod`/`sum`/`std`/`var`, where the
default is to compute the aggregation of the flattened array (e.g.,
`numpy.mean(arr_2d)` as opposed to `numpy.mean(arr_2d, axis=0)`).

`agg` is an alias for `aggregate`. Use the alias.

Examples
--------

>>> df = df = pd.DataFrame([[1,2,3],
...                         [4,5,6],
...                         [7,8,9],
...                         [np.nan, np.nan, np.nan]],
...                        columns=['A', 'B', 'C'])

Aggregate these functions across all columns

>>> df.aggregate(['sum', 'min'])
        A     B     C
sum  12.0  15.0  18.0
min   1.0   2.0   3.0

Different aggregations per column

>>> df.aggregate({'A' : ['sum', 'min'], 'B' : ['min', 'max']})
        A    B
max   NaN  8.0
min   1.0  2.0
sum  12.0  NaN

See also
--------
pandas.DataFrame.apply : Perform any type of operations.
pandas.DataFrame.transform : Perform transformation type operations.
pandas.DataFrame.groupby.aggregate : Perform aggregation type operations
    over groups.
pandas.DataFrame.resample.aggregate : Perform aggregation type operations
    over resampled bins.
pandas.DataFrame.rolling.aggregate : Perform aggregation type operations
    over rolling window.

################################################################################
################################## Validation ##################################
################################################################################

Errors found:
	Errors in parameters section
		Parameter "axis" description should start with capital letter
		Parameter "args" has no type
		Parameter "kwargs" has no type

```

If the validation script still gives errors, but you think there is a good reason
to deviate in this case (and there are certainly such cases), please state this
explicitly.

- args and kwargs have no type
- axis : {0 or 'index', 1 or 'columns'}, default 0

This PR has been made in collaboration with @rochamatcomp

Question to reviewers: we have added to the docstring the parameter `axis` because this is appropriate for DataFrame. However this dosctring is generic (shared with Series, groupby, window and resample) and the parameter `axis` is wrong for them (for Series the parameter has only one possible value). What should be the right way to fix this?

### CODE REVIEW

File: pandas/core/series.py [hunk 1]
```diff
@@ -2353,6 +2353,13 @@ def _gotitem(self, key, ndim, subset=None):
         return self
 
     _agg_doc = dedent("""
+    Notes
+    -----
```
Comment (TomAugspurger) [inline pandas/core/series.py]: Can remove this section. Move the axis discussion to a substitution in parameters.

File: pandas/core/frame.py [hunk 2]
```diff
@@ -4764,36 +4764,49 @@ def _gotitem(self, key, ndim, subset=None):
...
+    ...                         [np.nan, np.nan, np.nan]],
+    ...                        columns=['A', 'B', 'C'])
 
     Aggregate these functions across all columns
 
-    >>> df.agg(['sum', 'min'])
-                A         B         C
-    sum -0.182253 -0.614014 -2.909534
-    min -1.916563 -1.460076 -1.568297
+    >>> df.aggregate(['sum', 'min'])
```
Comment (TomAugspurger) [inline pandas/core/frame.py]: I think we recommend using `.agg`

File: pandas/core/frame.py [hunk 3]
```diff
@@ -4764,36 +4764,49 @@ def _gotitem(self, key, ndim, subset=None):
         return self[key]
 
     _agg_doc = dedent("""
+    Notes
+    -----
+    The default behavior of aggregating over the axis 0 is different from
+    `numpy` functions `mean`/`median`/`prod`/`sum`/`std`/`var`, where the
```
Comment (TomAugspurger) [inline pandas/core/frame.py]: Does sphinx complain about having a character immediately after the backticks? May need spaces.

File: pandas/core/frame.py [hunk 4]
```diff
@@ -4764,36 +4764,49 @@ def _gotitem(self, key, ndim, subset=None):
...
+    default is to compute the aggregation of the flattened array (e.g.,
+    `numpy.mean(arr_2d)` as opposed to `numpy.mean(arr_2d, axis=0)`).
```
Comment (TomAugspurger) [inline pandas/core/frame.py]: double-ticks around the code snippets.

File: pandas/core/frame.py [hunk 5]
```diff
@@ -4764,36 +4764,49 @@ def _gotitem(self, key, ndim, subset=None):
...
 
     See also
     --------
-    pandas.DataFrame.apply
-    pandas.DataFrame.transform
-    pandas.DataFrame.groupby.aggregate
-    pandas.DataFrame.resample.aggregate
-    pandas.DataFrame.rolling.aggregate
-
+    pandas.DataFrame.apply : Perform any type of operations.
```
Comment (TomAugspurger) [inline pandas/core/frame.py]: You can remove the `pandas.` from these, save a bit of space.

File: pandas/core/frame.py [hunk 6]
```diff
@@ -4764,36 +4764,49 @@ def _gotitem(self, key, ndim, subset=None):
...
+    pandas.DataFrame.transform : Perform transformation type operations.
+    pandas.DataFrame.groupby.aggregate : Perform aggregation type operations
```
Comment (TomAugspurger) [inline pandas/core/frame.py]: This should be pandas.core.groupby.GroupBy.transform.

Comment (TomAugspurger) [reply pandas/core/frame.py]: Actually, best to just remove the tranform / aggregate bits and just link to

pandas.core.groupby.GroupBy
pandas.core.resample.Resampler
pandas.core.window.Rolling
pandas.core.window.Expanding

File: pandas/core/generic.py [hunk 7]
```diff
@@ -3929,36 +3929,38 @@ def pipe(self, func, *args, **kwargs):
         return com._pipe(self, func, *args, **kwargs)
 
     _shared_docs['aggregate'] = ("""
-    Aggregate using callable, string, dict, or list of string/callables
+    Aggregate using one or multiple operations along the specified axis.
```
Comment (TomAugspurger) [inline pandas/core/generic.py]: multiple -> more

Comment (TomAugspurger) [reply pandas/core/generic.py]: along -> over I think.

Does along mean apply to each element in an axis? And over means collapse an axis?

**Review by TomAugspurger — CHANGES_REQUESTED** [review]
Regarding `axis`, I would add a variable in the shared doc `%(axis)s`.

In core/frame you can sub `axis : {0 or 'index', 1 or 'columns'}, default 0` and an explanation

In core/series you can sub `axis: {0 or 'index}` and note that it's for compatibility.

File: pandas/core/frame.py [hunk 8]
```diff
@@ -4764,36 +4764,49 @@ def _gotitem(self, key, ndim, subset=None):
...
 
     _agg_doc = dedent("""
+    Notes
+    -----
+    The default behavior of aggregating over the axis 0 is different from
+    `numpy` functions `mean`/`median`/`prod`/`sum`/`std`/`var`, where the
+    default is to compute the aggregation of the flattened array (e.g.,
+    `numpy.mean(arr_2d)` as opposed to `numpy.mean(arr_2d, axis=0)`).
+
+    `agg` is an alias for `aggregate`. Use the alias.
```
Comment (jreback) [inline pandas/core/frame.py]: I am not sure this is relevant, we *always* specify an axis, so by-definition you always have column-by-column or row-by-row aggregations, ``axis=None`` is simply not possible.

Comment (jreback) [reply pandas/core/frame.py]: maybe adding some expl would help here.

File: pandas/core/frame.py [hunk 9]
```diff
@@ -4764,36 +4764,49 @@ def _gotitem(self, key, ndim, subset=None):
...
+
     Examples
     --------
 
-    >>> df = pd.DataFrame(np.random.randn(10, 3), columns=['A', 'B', 'C'],
-    ...                   index=pd.date_range('1/1/2000', periods=10))
-    >>> df.iloc[3:7] = np.nan
+    >>> df = df = pd.DataFrame([[1,2,3],
```
Comment (jreback) [inline pandas/core/frame.py]: extra df =

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by albertvillanova** [comment]
I have read that there has been a change to the convention during the sprint, and `args` and `kwargs` should be in the same line and a common description in the next line. Should we implement this?

File: pandas/core/generic.py [hunk 10]
```diff
@@ -3929,36 +3929,36 @@ def pipe(self, func, *args, **kwargs):
...
-    mimicking the default Numpy behavior (e.g., np.mean(arr_2d)).
+        Accepted combinations are:
 
-    `agg` is an alias for `aggregate`. Use the alias.
+        - string function name.
+        - function.
+        - list of functions.
+        - dict of column names -> functions (or list of functions).
+    %(axis)s
+    args
```
Comment (TomAugspurger) [inline pandas/core/generic.py]: In this case, we'll document them separately.

This should be `*args`.

File: pandas/core/generic.py [hunk 11]
```diff
@@ -3929,36 +3929,36 @@ def pipe(self, func, *args, **kwargs):
...
+        Optional positional arguments to pass to the function.
```
Comment (TomAugspurger) [inline pandas/core/generic.py]: Strike optional. The agg func could have additional required parameters.

File: pandas/core/generic.py [hunk 12]
```diff
@@ -3929,36 +3929,36 @@ def pipe(self, func, *args, **kwargs):
...
+    kwargs
```
Comment (TomAugspurger) [inline pandas/core/generic.py]: This should be `**kwargs`.

**Review by TomAugspurger — COMMENTED** [review]
In this case we'll document them separately.

File: pandas/core/frame.py (hunk 9 above)
Comment (TomAugspurger) [reply pandas/core/frame.py]: Still have this @albertvillanova

File: pandas/core/frame.py (hunk 6 above)
Comment (albertvillanova) [reply pandas/core/frame.py]: @TomAugspurger  Also pandas.core.window.EWM?

**Comment by TomAugspurger** [comment]
Sure

File: pandas/core/frame.py (hunk 8 above)
Comment (albertvillanova) [reply pandas/core/frame.py]: Is it better now?

File: pandas/core/frame.py (hunk 3 above)
Comment (albertvillanova) [reply pandas/core/frame.py]: I have added commas instead.

**Comment by TomAugspurger** [comment]
Looks like there's a merge conflict @albertvillanova. Can you update? LMK if you need help with the git stuff.

**Review by TomAugspurger — APPROVED** [review]
Added an example with `axis='columns'`.

**Comment by albertvillanova** [comment]
Please, note that this PR has been made in collaboration with @rochamatcomp

**Pull request merged**